File: fastapi-backend/app/services/firebase_service.py
from datetime import datetime
from typing import Dict, Any, List, Optional
from app.core.firebase import db

class FirebaseService:
    """Service for Firebase operations"""
    
    @staticmethod
    async def get_canvas_credentials(uid: str) -> Dict[str, str]:
        """
        Get Canvas credentials for a user
        
        Args:
            uid: User ID
            
        Returns:
            Dict with Canvas URL and API key
        """
        try:
            user_doc = db.collection('users').document(uid).get()
            
            if not user_doc.exists:
                raise ValueError('User not found')
            
            user_data = user_doc.to_dict()
            
            if not user_data.get('canvasCredentials'):
                raise ValueError('Canvas credentials not found')
            
            return {
                "canvasUrl": user_data['canvasCredentials']['url'],
                "canvasApiKey": user_data['canvasCredentials']['apiKey']
            }
        except Exception as error:
            print(f'Error getting Canvas credentials: {error}')
            raise
    
    @staticmethod
    async def store_canvas_credentials(uid: str, credentials: Dict[str, str]) -> Dict[str, Any]:
        """
        Store Canvas credentials for a user
        
        Args:
            uid: User ID
            credentials: Canvas credentials (canvasUrl, canvasApiKey)
            
        Returns:
            Dict with operation result
        """
        try:
            canvas_url = credentials.get('canvasUrl')
            canvas_api_key = credentials.get('canvasApiKey')
            
            if not canvas_url or not canvas_api_key:
                raise ValueError('Canvas URL and API key are required')
            
            # Store credentials in Firestore
            db.collection('users').document(uid).set({
                'canvasCredentials': {
                    'url': canvas_url,
                    'apiKey': canvas_api_key,
                    'updatedAt': datetime.now()
                }
            }, merge=True)
            
            return {"success": True}
        except Exception as error:
            print(f'Error storing Canvas credentials: {error}')
            raise
    
    @staticmethod
    async def get_current_courses(uid: str) -> List[int]:
        """
        Get current course IDs from Firestore
        
        Args:
            uid: User ID
            
        Returns:
            List of current course IDs
        """
        try:
            courses_snapshot = db.collection('users').document(uid).collection('courses')\
                .where('status', '==', 'current').get()
            
            if not courses_snapshot:
                return []
            
            return [int(doc.id) for doc in courses_snapshot]
        except Exception as error:
            print(f'Error getting current courses: {error}')
            raise
    
    @staticmethod
    async def get_all_courses(uid: str) -> List[int]:
        """
        Get all course IDs from Firestore
        
        Args:
            uid: User ID
            
        Returns:
            List of all course IDs
        """
        try:
            courses_snapshot = db.collection('users').document(uid).collection('courses').get()
            
            if not courses_snapshot:
                return []
            
            return [int(doc.id) for doc in courses_snapshot]
        except Exception as error:
            print(f'Error getting all courses: {error}')
            raise

# Create service instance
firebase_service = FirebaseService()

File: expressjs-backend/services/firebaseService.js
const { db, auth } = require('../config/firebase');

/**
 * Service for Firebase operations
 */
const firebaseService = {
  /**
   * Store Canvas credentials for a user
   * @param {string} uid - User ID
   * @param {Object} credentials - Canvas credentials
   * @returns {Promise<Object>} Result of operation
   */
  storeCanvasCredentials: async (uid, credentials) => {
    try {
      const { canvasUrl, canvasApiKey } = credentials;
      
      if (!canvasUrl || !canvasApiKey) {
        throw new Error('Canvas URL and API key are required');
      }
      
      // Store credentials in Firestore
      await db.collection('users').doc(uid).set({
        canvasCredentials: {
          url: canvasUrl,
          apiKey: canvasApiKey,
          updatedAt: new Date()
        }
      }, { merge: true });
      
      return { success: true };
    } catch (error) {
      console.error('Error storing Canvas credentials:', error);
      throw error;
    }
  },
  
  /**
   * Get Canvas credentials for a user
   * @param {string} uid - User ID
   * @returns {Promise<Object>} Canvas credentials
   */
  getCanvasCredentials: async (uid) => {
    try {
      const userDoc = await db.collection('users').doc(uid).get();
      
      if (!userDoc.exists) {
        throw new Error('User not found');
      }
      
      const userData = userDoc.data();
      
      if (!userData.canvasCredentials) {
        throw new Error('Canvas credentials not found');
      }
      
      return {
        canvasUrl: userData.canvasCredentials.url,
        canvasApiKey: userData.canvasCredentials.apiKey
      };
    } catch (error) {
      console.error('Error getting Canvas credentials:', error);
      throw error;
    }
  },
  
  /**
   * Store user's courses in Firestore
   * @param {string} uid - User ID
   * @param {Array} courses - List of courses
   * @returns {Promise<Object>} Result of operation
   */
  storeCourses: async (uid, courses) => {
    try {
      // Get current date for labeling current/past courses
      const now = new Date();
      
      // Process courses to add status (current/past)
      const processedCourses = courses.map(course => {
        // Determine if course is current based on end_at date
        const endDate = course.end_at ? new Date(course.end_at) : null;
        const isCurrent = endDate ? endDate > now : true;
        
        return {
          id: course.id,
          name: course.name,
          course_code: course.course_code,
          term: course.term ? course.term.name : null,
          start_at: course.start_at,
          end_at: course.end_at,
          status: isCurrent ? 'current' : 'past',
          updatedAt: new Date()
        };
      });
      
      // Store courses in Firestore
      const batch = db.batch();
      
      // Create courses collection for user
      const userCoursesRef = db.collection('users').doc(uid).collection('courses');
      
      // Add each course to batch
      for (const course of processedCourses) {
        const courseRef = userCoursesRef.doc(course.id.toString());
        batch.set(courseRef, course);
      }
      
      // Commit batch
      await batch.commit();
      
      return { success: true, courseCount: processedCourses.length };
    } catch (error) {
      console.error('Error storing courses:', error);
      throw error;
    }
  },
  
  /**
   * Get user's courses from Firestore
   * @param {string} uid - User ID
   * @param {Object} options - Options for filtering courses
   * @returns {Promise<Array>} List of courses
   */
  getCourses: async (uid, options = {}) => {
    try {
      const { status } = options;
      
      let query = db.collection('users').doc(uid).collection('courses');
      
      // Filter by status if provided
      if (status) {
        query = query.where('status', '==', status);
      }
      
      const snapshot = await query.get();
      
      if (snapshot.empty) {
        return [];
      }
      
      // Convert snapshot to array of courses
      return snapshot.docs.map(doc => doc.data());
    } catch (error) {
      console.error('Error getting courses:', error);
      throw error;
    }
  },
  
  /**
   * Update course status (current/past) for all user's courses
   * @param {string} uid - User ID
   * @returns {Promise<Object>} Result of operation
   */
  updateCourseStatus: async (uid) => {
    try {
      // Get current date
      const now = new Date();
      
      // Get all user's courses
      const coursesSnapshot = await db.collection('users').doc(uid).collection('courses').get();
      
      if (coursesSnapshot.empty) {
        return { success: true, updated: 0 };
      }
      
      // Create batch for updates
      const batch = db.batch();
      let updatedCount = 0;
      
      // Update each course status
      coursesSnapshot.docs.forEach(doc => {
        const course = doc.data();
        const endDate = course.end_at ? new Date(course.end_at) : null;
        const isCurrent = endDate ? endDate > now : true;
        const newStatus = isCurrent ? 'current' : 'past';
        
        // Only update if status has changed
        if (course.status !== newStatus) {
          batch.update(doc.ref, { 
            status: newStatus,
            updatedAt: new Date()
          });
          updatedCount++;
        }
      });
      
      // Commit batch if there are updates
      if (updatedCount > 0) {
        await batch.commit();
      }
      
      return { success: true, updated: updatedCount };
    } catch (error) {
      console.error('Error updating course status:', error);
      throw error;
    }
  }
};

module.exports = firebaseService;
